File: backend/app/egr_client.py
"""Асинхронный клиент к API ЕГР РБ."""

import asyncio
from typing import Any
from urllib.parse import quote

import httpx

API_BASE = "http://egr.gov.by/api/v2/egr"
RETRIES = 6
RETRY_SLEEP = 1.5
TIMEOUT = 25.0
MAX_CONCURRENCY = 8

_semaphore = asyncio.Semaphore(MAX_CONCURRENCY)


async def call_api(method: str, *path_args: str) -> tuple[bool, Any]:
    segments = [method, *[str(a) for a in path_args]]
    url = API_BASE + "/" + "/".join(quote(s, safe="") for s in segments)
    last_err: str | None = None

    async with _semaphore:
        for _ in range(RETRIES):
            try:
                async with httpx.AsyncClient(timeout=TIMEOUT) as client:
                    resp = await client.get(
                        url,
                        headers={
                            "Accept": "application/json",
                            "User-Agent": "egr-lookup/3.0",
                        },
                    )
                    if resp.status_code == 204:
                        return True, None
                    if resp.status_code == 200 and not resp.content:
                        last_err = "пустой ответ от сервера"
                        await asyncio.sleep(RETRY_SLEEP)
                        continue
                    resp.raise_for_status()
                    return True, resp.json()
            except Exception as exc:  # noqa: BLE001
                last_err = str(exc)
                await asyncio.sleep(RETRY_SLEEP)

    return False, last_err or "неизвестная ошибка"


async def fetch_parallel(
    methods: list[dict[str, str]],
    *path_args: str,
) -> dict[str, dict[str, Any]]:
    async def one(key: str, method: str, title: str) -> tuple[str, dict[str, Any]]:
        ok, data = await call_api(method, *path_args)
        return key, {"ok": ok, "method": method, "title": title, "data": data}

    tasks = [one(m["key"], m["method"], m["title"]) for m in methods]
    pairs = await asyncio.gather(*tasks)
    return dict(pairs)

File: worker/src/egr_client.ts
const API_BASE = "http://egr.gov.by/api/v2/egr";
const RETRIES = 6;
const RETRY_SLEEP_MS = 1500;
const TIMEOUT_MS = 25_000;
const MAX_CONCURRENCY = 8;

let active = 0;
const queue: Array<() => void> = [];

function acquire(): Promise<void> {
  if (active < MAX_CONCURRENCY) {
    active += 1;
    return Promise.resolve();
  }
  return new Promise((resolve) => queue.push(resolve));
}

function release(): void {
  const next = queue.shift();
  if (next) next();
  else active -= 1;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function callApi(
  method: string,
  ...pathArgs: string[]
): Promise<[boolean, unknown]> {
  const segments = [method, ...pathArgs.map(String)];
  const url =
    API_BASE + "/" + segments.map((s) => encodeURIComponent(s)).join("/");
  let lastErr: string | null = null;

  await acquire();
  try {
    for (let i = 0; i < RETRIES; i++) {
      try {
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
        const resp = await fetch(url, {
          method: "GET",
          headers: {
            Accept: "application/json",
            "User-Agent": "egr-lookup/3.0",
          },
          signal: controller.signal,
        });
        clearTimeout(timer);

        if (resp.status === 204) return [true, null];
        if (resp.status === 200) {
          const text = await resp.text();
          if (!text) {
            lastErr = "пустой ответ от сервера";
            await sleep(RETRY_SLEEP_MS);
            continue;
          }
          return [true, JSON.parse(text)];
        }
        lastErr = `HTTP ${resp.status}`;
        await sleep(RETRY_SLEEP_MS);
      } catch (exc) {
        lastErr = exc instanceof Error ? exc.message : String(exc);
        await sleep(RETRY_SLEEP_MS);
      }
    }
  } finally {
    release();
  }

  return [false, lastErr ?? "неизвестная ошибка"];
}

export async function fetchParallel(
  methods: Array<{ key: string; method: string; title: string }>,
  ...pathArgs: string[]
): Promise<Record<string, { ok: boolean; method: string; title: string; data: unknown }>> {
  const pairs = await Promise.all(
    methods.map(async (m) => {
      const [ok, data] = await callApi(m.method, ...pathArgs);
      return [m.key, { ok, method: m.method, title: m.title, data }] as const;
    }),
  );
  return Object.fromEntries(pairs);
}
